server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/TabletMutationPrepAttempt.java

```diff
@@ -0,0 +1,124 @@
+import java.util.List;
+
+import org.apache.accumulo.core.constraints.Violations;
+import org.apache.accumulo.core.data.Mutation;
+
+public class TabletMutationPrepAttempt {
+
+  private final Violations violations = new Violations();
+
+  private CommitSession commitSession;
+  private List<Mutation> violators;
+  private List<Mutation> nonViolators;
+  private boolean attemptedTabletPrep;
+
+  /**
+   * Return true if an attempt was made to update the writes in progress for the tablet and to
+   * retrieve the commit session from the tablet's memory, or false otherwise.
+   */
+  public boolean attemptedTabletPrep() {
+    return attemptedTabletPrep;
+  }
+
+  /**
+   * Return true if this attempt has a non-null commit session. If null, it indicates that either
+   * there was never an attempt made to get the commit session from the tablet's memory, or that
+   * either the tablet has a closed state or the tablet's memory is unavailable.
+   */
+  public boolean hasCommitSession() {
+    return commitSession != null;
+  }
+
+  /**
+   * Retrieve the commit session. Possibly null.
+   */
+  public CommitSession getCommitSession() {
+    return commitSession;
+  }
+
+  /**
+   * Retrieve the constraint violations found across the mutations. Never null.
+   */
+  public Violations getViolations() {
+    return violations;
+  }
+
+  /**
+   * Return true if at least one constraint violation was found, or false otherwise.
+   */
+  public boolean hasViolations() {
+    return !violations.isEmpty();
+  }
+
+  /**
+   * Return the list of mutations that violated a constraint. Either null or a non-empty list.
+   */
+  public List<Mutation> getViolators() {
+    return violators;
+  }
+
+  /**
+   * Return the list of mutations that did not violate any constraints. Either null or a non-empty
+   * list.
+   */
+  public List<Mutation> getNonViolators() {
+    return nonViolators;
+  }
+
+  /**
+   * Return true if at least one mutation that did not violate any constraints was found.
+   */
+  public boolean hasNonViolators() {
+    return nonViolators != null && !nonViolators.isEmpty();
+  }
+
+  /**
+   * Set the commit session that was retrieved from the tablet's memory, and set
+   * {@link #attemptedTabletPrep} to true.
+   */
+  void setCommitSession(final CommitSession commitSession) {
+    this.commitSession = commitSession;
+    attemptedTabletPrep = true;
+  }
+
+  /**
+   * Add a mutation that violates a constraint, along with information about the constraints that it
+   * violated.
+   */
+  void addViolator(final Mutation mutation, final Violations mutationViolations) {
+    if (violators == null) {
+      violators = new ArrayList<>();
+    }
+    violators.add(mutation);
+    violations.add(mutationViolations);
+  }
+
+  /**
+   * Add a mutation that did not violate any constraints.
+   */
+  void addNonViolator(final Mutation mutation) {
+    if (nonViolators == null) {
+      nonViolators = new ArrayList<>();
```

Comment:
Same comment as above re: thread-safety and null-ness.

Reply:
See comment above.